Composer/packages/client/src/pages/design/VisualEditorHeader.tsx

```diff
@@ -116,29 +117,35 @@ const useBreadcrumbs = (projectId: string, pluginConfig?: PluginConfig) => {
   if (encodedFocused) {
     // we've linked to an action, so put that in too
     breadcrumbArray.push({
-      key: 'action-' + focusPath,
+      key: `${Types.Action}-${focusPath}`,
       label: getActionName(possibleAction, pluginConfig),
     });
   }
 
   const currentDialog = (dialogs.find(({ id }) => id === dialogId) ?? dialogs[0]) as DialogInfo;
 
-  const selectedActions = useMemo(() => {
-    const actionSelected = Array.isArray(visualEditorSelection) && visualEditorSelection.length > 0;
-    if (!actionSelected) return [];
-
-    const selectedActions = visualEditorSelection.map((id) => get(currentDialog?.content, id));
-
-    return selectedActions;
-  }, [visualEditorSelection, currentDialog?.content]);
-
-  if (selectedActions.length === 1 && selectedActions[0] != null) {
-    const action = selectedActions[0] as any;
-    const actionName = getActionName(action, pluginConfig);
-
-    breadcrumbArray = [...breadcrumbArray.slice(0, 2), { key: 'action-' + actionName, label: actionName }];
-  }
-
+  // get newest label for breadcrumbs
+  useEffect(() => {
+    if (currentDialog.content) {
+      breadcrumbArray = breadcrumbArray.map((b) => {
+        const type = b.key.charAt(0);
+        const name = b.key.substr(2);
+
+        switch (type) {
+          case Types.Dialog:
+            b.label = getFriendlyName(currentDialog.content);
+            break;
+          case Types.Trigger:
+            b.label = getFriendlyName(get(currentDialog.content, `triggers[${name}]`));
+            break;
+          case Types.Action:
+            b.label = getActionName(get(currentDialog.content, name));
+            break;
+        }
+        return b;
+      });
+    }
+  }, [currentDialog?.content]);
```

Comment:
This looks like an anti-pattern. Should `breadcrumbArray` be in state?

Reply:
@alanlong9278 `useMemo` should work

Reply:
Use `useMemo`